<!DOCTYPE html>
<html lang="en">

  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="Content-Language" content="en">

    <meta name="author" content="Ravin Poudel">
    <meta name="description" content="Here you can find approaches to:
i) Sampling a fasta file randomly with replacement ii) Example of creating a HMM profile for gene iii) Search genes predicted from a genome against HMM profiles  Tools/Program used:: prodigal
Prodigal (Prokaryotic Dynamic Programming Genefinding Algorithm) is a microbial (bacterial and archaeal) is a gene/ORF finding program.
hmmer
Hmmer provides tools for making probabilistic models of protein and DNA sequence domain families –called profile hidden Markov models, profile HMMs, or just profiles – and for using these profiles to annotate new sequences, to search sequence databases for additional homologs, and to make deep multiple sequence alignments.">
    <meta name="keywords" content="bioinformatics,genomics,personal">

    
      <script src="https://twemoji.maxcdn.com/v/latest/twemoji.min.js" crossorigin="anonymous"></script>
    

    <meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="Is ability to detect a gene a function of seqencing depth?"/>
<meta name="twitter:description" content="Here you can find approaches to:
i) Sampling a fasta file randomly with replacement ii) Example of creating a HMM profile for gene iii) Search genes predicted from a genome against HMM profiles  Tools/Program used:: prodigal
Prodigal (Prokaryotic Dynamic Programming Genefinding Algorithm) is a microbial (bacterial and archaeal) is a gene/ORF finding program.
hmmer
Hmmer provides tools for making probabilistic models of protein and DNA sequence domain families –called profile hidden Markov models, profile HMMs, or just profiles – and for using these profiles to annotate new sequences, to search sequence databases for additional homologs, and to make deep multiple sequence alignments."/>

    <meta property="og:title" content="Is ability to detect a gene a function of seqencing depth?" />
<meta property="og:description" content="Here you can find approaches to:
i) Sampling a fasta file randomly with replacement ii) Example of creating a HMM profile for gene iii) Search genes predicted from a genome against HMM profiles  Tools/Program used:: prodigal
Prodigal (Prokaryotic Dynamic Programming Genefinding Algorithm) is a microbial (bacterial and archaeal) is a gene/ORF finding program.
hmmer
Hmmer provides tools for making probabilistic models of protein and DNA sequence domain families –called profile hidden Markov models, profile HMMs, or just profiles – and for using these profiles to annotate new sequences, to search sequence databases for additional homologs, and to make deep multiple sequence alignments." />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://ravinpoudel.github.io/posts/seqencing-depth/" />
<meta property="article:published_time" content="2020-06-17T00:00:00+00:00" />
<meta property="article:modified_time" content="2020-06-17T00:00:00+00:00" />


    
      <base href="https://ravinpoudel.github.io/posts/seqencing-depth/">
    
    <title>
  Is ability to detect a gene a function of seqencing depth? · Ravin Poudel
</title>

    
      <link rel="canonical" href="https://ravinpoudel.github.io/posts/seqencing-depth/">
    

    <link href="https://fonts.googleapis.com/css?family=Lato:400,700%7CMerriweather:300,700%7CSource+Code+Pro:400,700&display=swap" rel="stylesheet">
    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.13.0/css/all.css" integrity="sha384-Bfad6CLCknfcloXFOyFnlgtENryhrpZCe29RTifKEixXQZ38WheV+i/6YWSzkz3V" crossorigin="anonymous">
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/normalize/8.0.1/normalize.min.css" integrity="sha256-l85OmPOjvil/SOvVt3HnSSjzF1TUMyT9eV0c2BzEGzU=" crossorigin="anonymous" />

    
      
      
      <link rel="stylesheet" href="/css/coder.min.3219ef62ae52679b7a9c19043171c3cd9f523628c2a65f3ef247ee18836bc90b.css" integrity="sha256-MhnvYq5SZ5t6nBkEMXHDzZ9SNijCpl8&#43;8kfuGINryQs=" crossorigin="anonymous" media="screen" />
    

    

    
      
        
        
        <link rel="stylesheet" href="/css/coder-dark.min.e78e80fc3a585a4d1c8fc7f58623b6ff852411e38431a9cd1792877ecaa160f6.css" integrity="sha256-546A/DpYWk0cj8f1hiO2/4UkEeOEManNF5KHfsqhYPY=" crossorigin="anonymous" media="screen" />
      
    

    

    

    <link rel="icon" type="image/png" href="https://ravinpoudel.github.io/images/favicon-32x32.png" sizes="32x32">
    <link rel="icon" type="image/png" href="https://ravinpoudel.github.io/images/favicon-16x16.png" sizes="16x16">

    <meta name="generator" content="Hugo 0.72.0" />
  </head>

  
  
    
  
  <body class="colorscheme-auto"
        onload=" twemoji.parse(document.body); "
  >
    <main class="wrapper">
      <nav class="navigation">
  <section class="container">
    <a class="navigation-title" href="/">
      Ravin Poudel
    </a>
    
    <input type="checkbox" id="menu-toggle" />
    <label class="menu-button float-right" for="menu-toggle"><i class="fas fa-bars"></i></label>
    <ul class="navigation-list">
      
        
          <li class="navigation-item">
            <a class="navigation-link" href="https://ravinpoudel.github.io/about/">About</a>
          </li>
        
          <li class="navigation-item">
            <a class="navigation-link" href="https://ravinpoudel.github.io/posts/">Blog</a>
          </li>
        
          <li class="navigation-item">
            <a class="navigation-link" href="https://ravinpoudel.github.io/publications/">Publications</a>
          </li>
        
          <li class="navigation-item">
            <a class="navigation-link" href="https://ravinpoudel.github.io/projects/">Projects</a>
          </li>
        
          <li class="navigation-item">
            <a class="navigation-link" href="https://ravinpoudel.github.io/contact/">Contact</a>
          </li>
        
          <li class="navigation-item">
            <a class="navigation-link" href="https://ravinpoudel.github.io/images/RavinPoudel.pdf">Resume</a>
          </li>
        
      
      
    </ul>
    
  </section>
</nav>


      <div class="content">
        
  <section class="container post">
    <article>
      <header>
        <div class="post-title">
          <h1 class="title">Is ability to detect a gene a function of seqencing depth?</h1>
        </div>
        <div class="post-meta">
          <div class="date">
            <span class="posted-on">
              <i class="fas fa-calendar"></i>
              <time datetime='2020-06-17T00:00:00Z'>
                June 17, 2020
              </time>
            </span>
            <span class="reading-time">
              <i class="fas fa-clock"></i>
              4-minute read
            </span>
          </div>
          
          <div class="tags">
  <i class="fas fa-tag"></i>
    <a href="/tags/python/">Python</a>
      <span class="separator">•</span>
    <a href="/tags/bioinformatics/">Bioinformatics</a>
      <span class="separator">•</span>
    <a href="/tags/biopython/">Biopython</a></div>

        </div>
      </header>

      <div>
        
        <p>Here you can find approaches to:</p>
<pre><code>i)      Sampling a fasta file randomly with replacement
ii)     Example of creating a HMM profile for gene
iii)    Search genes predicted from a genome against HMM profiles
</code></pre>
<h5 id="toolsprogram-used">Tools/Program used::</h5>
<p><a href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-11-119">prodigal</a></p>
<p>Prodigal (Prokaryotic Dynamic Programming Genefinding Algorithm) is a microbial (bacterial and archaeal) is a gene/ORF finding program.</p>
<p><a href="http://hmmer.org">hmmer</a></p>
<p>Hmmer provides tools for making probabilistic models of protein and DNA sequence domain families –called profile hidden Markov models, profile HMMs, or just profiles – and for using these profiles to annotate new sequences, to search sequence databases for additional homologs, and to make deep multiple sequence alignments.
Steps:</p>
<ol>
<li>Multiple sequence alignment of a homologous family of protein</li>
<li>Making HMM models using the aligned output</li>
<li>Use the constructed HMM profile to annotate other sequence/aa.</li>
</ol>
<p><a href="https://www.ebi.ac.uk/Tools/msa/muscle/">muscle</a></p>
<p>MUSCLE (MUltiple Sequence Comparison by Log-Expectatio) is multiple alignment programs.</p>
<p>##############################</p>
<p>This code allows to randomly sample a large fasta file at n number of times with replacement. Once sampled, then gene is predicted for each sub-sampled fasta files using <code>prodigal</code>. Predicted gene file is searched for the gene of interest using HMM profile in <code>hmmer</code>. In order to run hmmer search,  at first protein sequnces for a gene of interest is downloaded from uniport, aligned with <code>muscle</code>, and HMM profile is created with <code>hmmbuild</code> function from <code>Hmmer</code>. The built HMM profile is then used to search gene of interest in the sub-sampled fasta files.</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-python" data-lang="python">
<span style="color:#75715e">#!/usr/bin/env python3</span>
<span style="color:#75715e"># -*- coding: utf-8 -*-</span>
<span style="color:#e6db74">&#34;&#34;&#34;
</span><span style="color:#e6db74">Created on Wed Jun 10 17:03:23 2020
</span><span style="color:#e6db74">
</span><span style="color:#e6db74">@author: ravinpoudel
</span><span style="color:#e6db74">&#34;&#34;&#34;</span>

<span style="color:#f92672">import</span> os
<span style="color:#f92672">import</span> subprocess
<span style="color:#f92672">import</span> pandas <span style="color:#f92672">as</span> pd
<span style="color:#f92672">import</span> matplotlib.pyplot <span style="color:#f92672">as</span> plt
<span style="color:#f92672">from</span> Bio <span style="color:#f92672">import</span> SeqIO
<span style="color:#f92672">from</span> random <span style="color:#f92672">import</span> choices
<span style="color:#f92672">from</span> random <span style="color:#f92672">import</span> seed

<span style="color:#75715e"># input fasta file</span>
fasta <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;MSM9VZPH_contigs.fna&#34;</span>


<span style="color:#66d9ef">def</span> <span style="color:#a6e22e">rsamplefasta</span>(fasta, n, nseed):
    
    <span style="color:#e6db74">&#34;&#34;&#34;Randomly sample n number of reads - with replacement&#34;&#34;&#34;</span>
    
    <span style="color:#75715e"># fname to write as an output filename</span>
    fname <span style="color:#f92672">=</span> os<span style="color:#f92672">.</span>path<span style="color:#f92672">.</span>splitext(fasta)[<span style="color:#ae81ff">0</span>] <span style="color:#f92672">+</span><span style="color:#e6db74">&#34;_&#34;</span><span style="color:#f92672">+</span> str(n) <span style="color:#f92672">+</span><span style="color:#e6db74">&#34;.fasta&#34;</span>
    
    <span style="color:#75715e"># index fasta file, allows for fast access</span>
    record_dict <span style="color:#f92672">=</span> SeqIO<span style="color:#f92672">.</span>index(fasta, <span style="color:#e6db74">&#34;fasta&#34;</span>)
    
    <span style="color:#75715e"># create a list of sequence header</span>
    seqheader <span style="color:#f92672">=</span> []
    <span style="color:#66d9ef">for</span> file <span style="color:#f92672">in</span> record_dict<span style="color:#f92672">.</span>keys():
        seqheader<span style="color:#f92672">.</span>append(file)
    
    <span style="color:#75715e"># randomly sequence for seqheader  for &#39;n&#39; times</span>
    seed(nseed)
    sub_header <span style="color:#f92672">=</span> choices(seqheader, k<span style="color:#f92672">=</span>n)
    
    <span style="color:#75715e"># retrive reads/seqrecord from fasta file, based on header in seqheader list</span>
    sample_fasta <span style="color:#f92672">=</span> []
    <span style="color:#66d9ef">for</span> header <span style="color:#f92672">in</span> sub_header:
        sample_fasta<span style="color:#f92672">.</span>append(record_dict[header])
 
    <span style="color:#75715e"># write down as a fasta file</span>
    SeqIO<span style="color:#f92672">.</span>write(sample_fasta, fname , <span style="color:#e6db74">&#34;fasta&#34;</span>)
    
    
<span style="color:#66d9ef">def</span> <span style="color:#a6e22e">rangesampler_fasta</span>(fasta):
    <span style="color:#e6db74">&#34;&#34;&#34;Randomly sample [1000, 10000, 100000, 1000000] reads - with replacement&#34;&#34;&#34;</span>
    
    <span style="color:#75715e"># sample range</span>
    samplerange <span style="color:#f92672">=</span> [<span style="color:#ae81ff">1000</span>, <span style="color:#ae81ff">10000</span>, <span style="color:#ae81ff">100000</span>, <span style="color:#ae81ff">1000000</span>]
    
    <span style="color:#75715e"># random seed</span>
    nseed <span style="color:#f92672">=</span> <span style="color:#ae81ff">12121232</span>
    
    <span style="color:#75715e"># sample randomly n times and save as a fasta file</span>
    <span style="color:#66d9ef">for</span> i <span style="color:#f92672">in</span> samplerange:
        <span style="color:#66d9ef">print</span>(<span style="color:#e6db74">&#34;Ranomly Sampling&#34;</span>, i, <span style="color:#e6db74">&#34;sequences&#34;</span>)
        rsamplefasta(fasta, i, nseed)
        
        
<span style="color:#75715e">#rsamplefasta(fasta, 20, nseed)</span>

<span style="color:#75715e"># Randomly sample reads from a given fasta file @ n = 1000, 10000, 100000, 1000000</span>
rangesampler_fasta(fasta)


<span style="color:#75715e"># make a gene prediction using prodigal , for each n sampled fasta files</span>
subprocess<span style="color:#f92672">.</span>call(<span style="color:#e6db74">&#34;./prodigal.sh&#34;</span>)
</code></pre></div><p><code>prodigal.sh</code></p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-bash" data-lang="bash"><span style="color:#75715e">#!/bin/bash
</span><span style="color:#75715e"></span><span style="color:#66d9ef">for</span> file in <span style="color:#66d9ef">$(</span> ls *_contigs_1*.fasta<span style="color:#66d9ef">)</span>;<span style="color:#66d9ef">do</span>
     echo file: $file
     fname<span style="color:#f92672">=</span><span style="color:#66d9ef">$(</span>basename $file .fasta<span style="color:#66d9ef">)</span>
     echo basename: $fname
     time prodigal -p meta -a $fname.faa -i $file
<span style="color:#66d9ef">done</span>
</code></pre></div><div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-python" data-lang="python"><span style="color:#75715e">## Run hmmer search - with gene of interest - eg. bft here. </span>
<span style="color:#75715e"># basic order: gene faa &gt; align with mafft &gt; hmm profile &gt; hmmsearch </span>
subprocess<span style="color:#f92672">.</span>call(<span style="color:#e6db74">&#34;./alignthenHmmer.sh&#34;</span>)

</code></pre></div><p><code>alignthenHmmer.sh</code></p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-bash" data-lang="bash"><span style="color:#75715e">#!/bin/bash
</span><span style="color:#75715e"></span><span style="color:#75715e"># align bft faa file.</span> 
muscle -in bft.faa -out bft_aln.faa 

<span style="color:#75715e"># use hmmbuild to create a hmm profile with align bft file</span>
hmmbuild bft.hmm bft_aln.faa 

<span style="color:#75715e"># hmmsearch of metagenome.faa with hmm profile</span>
<span style="color:#66d9ef">for</span> file in <span style="color:#66d9ef">$(</span> ls *_contigs_1*.faa<span style="color:#66d9ef">)</span>;<span style="color:#66d9ef">do</span>
     echo file: $file
     fname<span style="color:#f92672">=</span><span style="color:#66d9ef">$(</span>basename $file .faa<span style="color:#66d9ef">)</span>
     echo basename: $fname
     time hmmsearch --tblout $fname.bft.txt -E 1e-6 --cpu <span style="color:#ae81ff">16</span> bft.hmm $file
<span style="color:#66d9ef">done</span>

</code></pre></div><div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-python" data-lang="python">
<span style="color:#75715e">## Now count the number of hits for each file</span>
subprocess<span style="color:#f92672">.</span>call(<span style="color:#e6db74">&#34;./counthits.sh&#34;</span>)

</code></pre></div><p><code>counthits.sh</code></p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-bash" data-lang="bash"><span style="color:#75715e">#!/bin/bash
</span><span style="color:#75715e"></span><span style="color:#66d9ef">for</span> file in <span style="color:#66d9ef">$(</span> ls *_<span style="color:#f92672">[</span>0-9<span style="color:#f92672">]</span>*.bft.txt<span style="color:#66d9ef">)</span>;<span style="color:#66d9ef">do</span>
     echo file: $file
     fname<span style="color:#f92672">=</span><span style="color:#66d9ef">$(</span>basename $file .bft.txt<span style="color:#66d9ef">)</span>
     echo basename: $fname
     nline<span style="color:#f92672">=</span><span style="color:#66d9ef">$(</span>grep -vc <span style="color:#e6db74">&#39;^#&#39;</span> $file<span style="color:#66d9ef">)</span>
     echo $file $nline &gt;&gt; countline.txt
<span style="color:#66d9ef">done</span>

</code></pre></div><div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-python" data-lang="python"><span style="color:#75715e">### count the number of matches </span>
lcount <span style="color:#f92672">=</span> pd<span style="color:#f92672">.</span>read_csv(<span style="color:#e6db74">&#34;countline.txt&#34;</span>, header<span style="color:#f92672">=</span>None, sep<span style="color:#f92672">=</span><span style="color:#e6db74">&#34; &#34;</span>, names<span style="color:#f92672">=</span>[<span style="color:#e6db74">&#39;Filename&#39;</span>,<span style="color:#e6db74">&#39;nhits&#39;</span>])
lcount[<span style="color:#e6db74">&#39;nreads&#39;</span>] <span style="color:#f92672">=</span> lcount<span style="color:#f92672">.</span>Filename<span style="color:#f92672">.</span>str<span style="color:#f92672">.</span>rsplit(<span style="color:#e6db74">&#34;_&#34;</span>, n<span style="color:#f92672">=</span><span style="color:#ae81ff">1</span>, expand<span style="color:#f92672">=</span>True)[<span style="color:#ae81ff">1</span>]<span style="color:#f92672">.</span>str<span style="color:#f92672">.</span>rsplit(<span style="color:#e6db74">&#34;.bft.txt&#34;</span>, expand<span style="color:#f92672">=</span>True)[<span style="color:#ae81ff">0</span>]<span style="color:#f92672">.</span>to_list()


<span style="color:#75715e">### Create a plot</span>

plt<span style="color:#f92672">.</span>plot(lcount<span style="color:#f92672">.</span>nreads<span style="color:#f92672">.</span>to_list(), lcount<span style="color:#f92672">.</span>nhits<span style="color:#f92672">.</span>to_list())
plt<span style="color:#f92672">.</span>ylabel(<span style="color:#e6db74">&#39;Number of hits for bft gene&#39;</span>)
plt<span style="color:#f92672">.</span>xlabel(<span style="color:#e6db74">&#34;Number of reads sampled - with replacement&#34;</span>)
<span style="color:#75715e">#plt.grid(True)</span>
plt<span style="color:#f92672">.</span>savefig(<span style="color:#e6db74">&#39;bft.png&#39;</span>, dpi<span style="color:#f92672">=</span><span style="color:#ae81ff">300</span>)
plt<span style="color:#f92672">.</span>show()  

<span style="color:#75715e">#####</span>
</code></pre></div><p><img src="/images/bft.png" alt="alt text"></p>

      </div>


      <footer>
        


        <div id="disqus_thread"></div>
<script type="application/javascript">
    var disqus_config = function () {
    
    
    
    };
    (function() {
        if (["localhost", "127.0.0.1"].indexOf(window.location.hostname) != -1) {
            document.getElementById('disqus_thread').innerHTML = 'Disqus comments not available by default when the website is previewed locally.';
            return;
        }
        var d = document, s = d.createElement('script'); s.async = true;
        s.src = '//' + "https-ravinpoudel-github-io" + '.disqus.com/embed.js';
        s.setAttribute('data-timestamp', +new Date());
        (d.head || d.body).appendChild(s);
    })();
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
<a href="https://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
        
        
      </footer>
    </article>

    
  </section>

      </div>

      
  <footer class="footer">
    <section class="container">
      
      
        ©
        
          2000 -
        
        2020
         Ravin Poudel 
      
      
         · 
        Powered by <a href="https://gohugo.io/">Hugo</a> & <a href="https://github.com/luizdepra/hugo-coder/">Coder</a>.
      
      
    </section>
  </footer>

    </main>

    

    
<script type="application/javascript">
var doNotTrack = false;
if (!doNotTrack) {
	(function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
	(i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
	m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
	})(window,document,'script','https://www.google-analytics.com/analytics.js','ga');
	ga('create', 'UA-99029062-1', 'auto');
	
	ga('send', 'pageview');
}
</script>


    

  </body>

</html>
