<!DOCTYPE html>
<html lang="en">

  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="Content-Language" content="en">

    <meta name="author" content="Ravin Poudel">
    <meta name="description" content="Question/Task:: Clinical Genomics Design a targeted panel for sequencing all coding exons of human genes BRCA1 and BRCA2 (HGNC names given), and provide genomic coordinate ranges (as a BED- format file) that encompass all bases in those coding exons to the panel manufacturer. The ranges must be disjoint (no overlap, and at least one genomic base must separate adjacent intervals). Such output will be used by the manufacturer to design PCR primer pairs that amplify these targets out of genomic DNA for downstream NGS.">
    <meta name="keywords" content="bioinformatics,genomics,personal">

    
      <script src="https://twemoji.maxcdn.com/v/latest/twemoji.min.js" crossorigin="anonymous"></script>
    

    <meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="Designing a targeted panel - for Clinical Genomics"/>
<meta name="twitter:description" content="Question/Task:: Clinical Genomics Design a targeted panel for sequencing all coding exons of human genes BRCA1 and BRCA2 (HGNC names given), and provide genomic coordinate ranges (as a BED- format file) that encompass all bases in those coding exons to the panel manufacturer. The ranges must be disjoint (no overlap, and at least one genomic base must separate adjacent intervals). Such output will be used by the manufacturer to design PCR primer pairs that amplify these targets out of genomic DNA for downstream NGS."/>

    <meta property="og:title" content="Designing a targeted panel - for Clinical Genomics" />
<meta property="og:description" content="Question/Task:: Clinical Genomics Design a targeted panel for sequencing all coding exons of human genes BRCA1 and BRCA2 (HGNC names given), and provide genomic coordinate ranges (as a BED- format file) that encompass all bases in those coding exons to the panel manufacturer. The ranges must be disjoint (no overlap, and at least one genomic base must separate adjacent intervals). Such output will be used by the manufacturer to design PCR primer pairs that amplify these targets out of genomic DNA for downstream NGS." />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://ravinpoudel.github.io/posts/exome_gene_panel/" />
<meta property="article:published_time" content="2020-05-30T00:00:00+00:00" />
<meta property="article:modified_time" content="2020-05-30T00:00:00+00:00" />


    
      <base href="https://ravinpoudel.github.io/posts/exome_gene_panel/">
    
    <title>
  Designing a targeted panel - for Clinical Genomics · Ravin Poudel
</title>

    
      <link rel="canonical" href="https://ravinpoudel.github.io/posts/exome_gene_panel/">
    

    <link href="https://fonts.googleapis.com/css?family=Lato:400,700%7CMerriweather:300,700%7CSource+Code+Pro:400,700&display=swap" rel="stylesheet">
    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.13.0/css/all.css" integrity="sha384-Bfad6CLCknfcloXFOyFnlgtENryhrpZCe29RTifKEixXQZ38WheV+i/6YWSzkz3V" crossorigin="anonymous">
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/normalize/8.0.1/normalize.min.css" integrity="sha256-l85OmPOjvil/SOvVt3HnSSjzF1TUMyT9eV0c2BzEGzU=" crossorigin="anonymous" />

    
      
      
      <link rel="stylesheet" href="/css/coder.min.3219ef62ae52679b7a9c19043171c3cd9f523628c2a65f3ef247ee18836bc90b.css" integrity="sha256-MhnvYq5SZ5t6nBkEMXHDzZ9SNijCpl8&#43;8kfuGINryQs=" crossorigin="anonymous" media="screen" />
    

    

    
      
        
        
        <link rel="stylesheet" href="/css/coder-dark.min.e78e80fc3a585a4d1c8fc7f58623b6ff852411e38431a9cd1792877ecaa160f6.css" integrity="sha256-546A/DpYWk0cj8f1hiO2/4UkEeOEManNF5KHfsqhYPY=" crossorigin="anonymous" media="screen" />
      
    

    

    

    <link rel="icon" type="image/png" href="https://ravinpoudel.github.io/images/favicon-32x32.png" sizes="32x32">
    <link rel="icon" type="image/png" href="https://ravinpoudel.github.io/images/favicon-16x16.png" sizes="16x16">

    <meta name="generator" content="Hugo 0.72.0" />
  </head>

  
  
    
  
  <body class="colorscheme-auto"
        onload=" twemoji.parse(document.body); "
  >
    <main class="wrapper">
      <nav class="navigation">
  <section class="container">
    <a class="navigation-title" href="/">
      Ravin Poudel
    </a>
    
    <input type="checkbox" id="menu-toggle" />
    <label class="menu-button float-right" for="menu-toggle"><i class="fas fa-bars"></i></label>
    <ul class="navigation-list">
      
        
          <li class="navigation-item">
            <a class="navigation-link" href="https://ravinpoudel.github.io/about/">About</a>
          </li>
        
          <li class="navigation-item">
            <a class="navigation-link" href="https://ravinpoudel.github.io/posts/">Blog</a>
          </li>
        
          <li class="navigation-item">
            <a class="navigation-link" href="https://ravinpoudel.github.io/publications/">Publications</a>
          </li>
        
          <li class="navigation-item">
            <a class="navigation-link" href="https://ravinpoudel.github.io/projects/">Projects</a>
          </li>
        
          <li class="navigation-item">
            <a class="navigation-link" href="https://ravinpoudel.github.io/contact/">Contact</a>
          </li>
        
          <li class="navigation-item">
            <a class="navigation-link" href="https://ravinpoudel.github.io/images/RavinPoudel.pdf">Resume</a>
          </li>
        
      
      
    </ul>
    
  </section>
</nav>


      <div class="content">
        
  <section class="container post">
    <article>
      <header>
        <div class="post-title">
          <h1 class="title">Designing a targeted panel - for Clinical Genomics</h1>
        </div>
        <div class="post-meta">
          <div class="date">
            <span class="posted-on">
              <i class="fas fa-calendar"></i>
              <time datetime='2020-05-30T00:00:00Z'>
                May 30, 2020
              </time>
            </span>
            <span class="reading-time">
              <i class="fas fa-clock"></i>
              2-minute read
            </span>
          </div>
          
          <div class="tags">
  <i class="fas fa-tag"></i>
    <a href="/tags/python/">Python</a>
      <span class="separator">•</span>
    <a href="/tags/bioinformatics/">Bioinformatics</a>
      <span class="separator">•</span>
    <a href="/tags/biopython/">Biopython</a></div>

        </div>
      </header>

      <div>
        
        <h2 id="questiontask-clinical-genomics">Question/Task:: Clinical Genomics</h2>
<p>Design a targeted panel for sequencing all coding exons of human genes BRCA1 and BRCA2 (HGNC names given), and provide genomic coordinate ranges (as a BED- format file) that encompass all bases in those coding exons to the panel manufacturer. The ranges must be disjoint (no overlap, and at least one genomic base must separate adjacent intervals). Such output will be used by the manufacturer to design PCR primer pairs that amplify these targets out of genomic DNA for downstream NGS.</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-python" data-lang="python"><span style="color:#75715e">#!/usr/bin/env python3</span>
<span style="color:#75715e"># -*- coding: utf-8 -*-</span>
<span style="color:#e6db74">&#34;&#34;&#34;
</span><span style="color:#e6db74">Created on Fri May 22 18:15:20 2020
</span><span style="color:#e6db74">
</span><span style="color:#e6db74">@author: ravinpoudel
</span><span style="color:#e6db74">&#34;&#34;&#34;</span>

<span style="color:#f92672">import</span> pyfaidx
<span style="color:#f92672">import</span> pandas <span style="color:#f92672">as</span> pd
<span style="color:#f92672">import</span> pybedtools
<span style="color:#f92672">from</span> collections <span style="color:#f92672">import</span> deque
<span style="color:#f92672">from</span> pyensembl <span style="color:#f92672">import</span> EnsemblRelease



<span style="color:#75715e"># retrive HGNC - release 77 uses human reference genome GRCh38</span>
ensembldata <span style="color:#f92672">=</span> EnsemblRelease(<span style="color:#ae81ff">77</span>)

<span style="color:#75715e"># index human genome. Indexning allows for fasta random access to the genome.</span>
<span style="color:#75715e"># this will speed things when we need to retrive actual nucleotide sequnce for gene cordinates</span>
refidx <span style="color:#f92672">=</span> pyfaidx<span style="color:#f92672">.</span>Faidx(<span style="color:#e6db74">&#39;reference/Homo_sapiens.GRCh38.dna.toplevel.fa&#39;</span>)
genename <span style="color:#f92672">=</span><span style="color:#e6db74">&#39;BRCA1&#39;</span>



<span style="color:#66d9ef">def</span> <span style="color:#a6e22e">getbed4exon</span>(ensembldata, refidx, genename):
<span style="color:#e6db74">&#34;&#34;&#34;Provide non-overlapping genomic coordinate ranges (as a BED- format file with and without sequnces)
</span><span style="color:#e6db74">for a gene &#34;&#34;&#34;</span>
    
    <span style="color:#75715e">## get exon ids for geneid</span>
    exon_ids <span style="color:#f92672">=</span> ensembldata<span style="color:#f92672">.</span>exon_ids_of_gene_name(genename)
    
    <span style="color:#75715e"># Get locus for each exon ids</span>
    exon_locus <span style="color:#f92672">=</span> deque()
    <span style="color:#66d9ef">for</span> i <span style="color:#f92672">in</span> exon_ids:
        record <span style="color:#f92672">=</span> ensembldata<span style="color:#f92672">.</span>locus_of_exon_id(i)
        exn <span style="color:#f92672">=</span> {<span style="color:#e6db74">&#39;contig&#39;</span> : record<span style="color:#f92672">.</span>contig,
                   <span style="color:#e6db74">&#39;start&#39;</span> : record<span style="color:#f92672">.</span>start,
                   <span style="color:#e6db74">&#39;end&#39;</span> : record<span style="color:#f92672">.</span>end,
                   <span style="color:#e6db74">&#39;length&#39;</span> : record<span style="color:#f92672">.</span>length,
                   <span style="color:#e6db74">&#39;strand&#39;</span> : record<span style="color:#f92672">.</span>strand}
        exon_locus<span style="color:#f92672">.</span>append(exn)
    
    <span style="color:#75715e"># list to datafram</span>
    df <span style="color:#f92672">=</span> pd<span style="color:#f92672">.</span>DataFrame(exon_locus)
    
    <span style="color:#75715e"># covert to bed then sort</span>
    df_bed <span style="color:#f92672">=</span> pybedtools<span style="color:#f92672">.</span>BedTool<span style="color:#f92672">.</span>from_dataframe(df)<span style="color:#f92672">.</span>sort()
    
    <span style="color:#75715e">## remove overlapping ranges</span>
    df_bed_merged <span style="color:#f92672">=</span> df_bed<span style="color:#f92672">.</span>merge()
    
    <span style="color:#75715e">## add sequence to the exon_locus</span>
    ex_seq <span style="color:#f92672">=</span> deque()
    <span style="color:#66d9ef">for</span> i <span style="color:#f92672">in</span> df_bed_merged:
        contig <span style="color:#f92672">=</span> i[<span style="color:#ae81ff">0</span>] 
        start <span style="color:#f92672">=</span> int(i[<span style="color:#ae81ff">1</span>])
        end <span style="color:#f92672">=</span> int(i[<span style="color:#ae81ff">2</span>])
        length <span style="color:#f92672">=</span> (end <span style="color:#f92672">-</span> start) <span style="color:#f92672">+</span> <span style="color:#ae81ff">1</span>
        exon_seq <span style="color:#f92672">=</span> str(fa<span style="color:#f92672">.</span>fetch(contig, start, end))
        record <span style="color:#f92672">=</span> {<span style="color:#e6db74">&#39;contig&#39;</span>: contig,
                   <span style="color:#e6db74">&#39;start&#39;</span>:  start,
                   <span style="color:#e6db74">&#39;end&#39;</span>: end,
                   <span style="color:#e6db74">&#39;seq&#39;</span>: exon_seq,
                   <span style="color:#e6db74">&#39;length&#39;</span>: length,
                   <span style="color:#e6db74">&#39;gene&#39;</span>: genename}
        ex_seq<span style="color:#f92672">.</span>append(record)
    
    <span style="color:#75715e">## export as bed file</span>
    final_exon <span style="color:#f92672">=</span> pd<span style="color:#f92672">.</span>DataFrame(ex_seq)
    final_bed <span style="color:#f92672">=</span> pybedtools<span style="color:#f92672">.</span>BedTool<span style="color:#f92672">.</span>from_dataframe(final_exon)
    <span style="color:#75715e"># final_bed.saveas(&#39;final.bed&#39;)</span>
    <span style="color:#66d9ef">print</span>(<span style="color:#e6db74">&#39;Number of unoverlapping record in the bedfile for&#39;</span>, genename,<span style="color:#e6db74">&#39;::&#39;</span>, len(final_bed))
    <span style="color:#66d9ef">return</span> (df_bed_merged, final_bed)

<span style="color:#75715e">##</span>

bed_brca1_cordinates, bed_brca1_cordinates_withsequence <span style="color:#f92672">=</span> getbed4exon(ensembldata, refidx, <span style="color:#e6db74">&#39;BRCA1&#39;</span>)
bed_brca2_cordinates, bed_brca2_cordinates_withsequence <span style="color:#f92672">=</span> <span style="color:#f92672">=</span> getbed4exon(ensembldata, refidx, <span style="color:#e6db74">&#39;BRCA2&#39;</span>)


</code></pre></div>
      </div>


      <footer>
        


        <div id="disqus_thread"></div>
<script type="application/javascript">
    var disqus_config = function () {
    
    
    
    };
    (function() {
        if (["localhost", "127.0.0.1"].indexOf(window.location.hostname) != -1) {
            document.getElementById('disqus_thread').innerHTML = 'Disqus comments not available by default when the website is previewed locally.';
            return;
        }
        var d = document, s = d.createElement('script'); s.async = true;
        s.src = '//' + "https-ravinpoudel-github-io" + '.disqus.com/embed.js';
        s.setAttribute('data-timestamp', +new Date());
        (d.head || d.body).appendChild(s);
    })();
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
<a href="https://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
        
        
      </footer>
    </article>

    
  </section>

      </div>

      
  <footer class="footer">
    <section class="container">
      
      
        ©
        
          2000 -
        
        2020
         Ravin Poudel 
      
      
         · 
        Powered by <a href="https://gohugo.io/">Hugo</a> & <a href="https://github.com/luizdepra/hugo-coder/">Coder</a>.
      
      
    </section>
  </footer>

    </main>

    

    
<script type="application/javascript">
var doNotTrack = false;
if (!doNotTrack) {
	(function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
	(i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
	m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
	})(window,document,'script','https://www.google-analytics.com/analytics.js','ga');
	ga('create', 'UA-99029062-1', 'auto');
	
	ga('send', 'pageview');
}
</script>


    

  </body>

</html>
